R/tbl_listing.R

```diff
@@ -0,0 +1,205 @@
+#' Create listings from a data frame
+#'
+#' @param df (`data.frame`)\cr
+#'   a data frame containing the data to be displayed in the listing.
+#' @param key_cols ([`tidy-select`][dplyr::dplyr_tidy_select])\cr
+#'   column names to be used as key columns in the listing.
+#' @param sort_cols ([`tidy-select`][dplyr::dplyr_tidy_select])\cr
+#'   column names to be used for sorting the data frame before creating the listing. It defaults to `key_cols`.
+#' @param add_blank_key_duplicates (`logical`)\cr
+#'   whether to add blank values where key columns have duplicate values. Defaults to `TRUE`.
+#' @param ... other objects that will be added to the `{gtsummary}` object list.
+#'
+#' @details
+#' This function converts a data frame into a listing format and it relies on the user's pre-processing
+#' to decide which column to show and select unique or non-unique rows. Also sorting depends on the user
+#' pre-processing.
+#'
+#' @examples
+#' # Example 1 --------------------------------
+#' trial_df <- trial |>
+#'   dplyr::select(trt, age, marker, stage)
+#'
+#' tbl_listing(trial_df, key_cols = c(trt, stage))
+#'
+#' @export
+tbl_listing <- function(df,
+                        key_cols = NULL,
+                        sort_cols = NULL,
+                        add_blank_key_duplicates = TRUE,
+                        ...) {
+  # Checks -----------------------------------
+  check_not_missing(df)
+  check_data_frame(df)
+  check_scalar_logical(add_blank_key_duplicates)
+
+  row_all_na <- apply(df, 1, function(x) all(is.na(x)))
+  if (any(row_all_na)) {
+    cli::cli_warn("Rows that only contain NA values have been trimmed.")
+    df <- df[!row_all_na, ]
+  }
+
+  # Explicit NAs -----------------------------
+  df <- df |>
+    mutate(across(everything(), ~ replace_na(as.character(.), "NA")))
```

Comment:
Here the NAs would be blank and we do not want that if not for keycols. we can provide access to this value. I am sure that this can be managed by fmt_fun, but I kept it simple for now

Reply:
na_explicit_fmt?

Reply:
Users's responsibility -> maybe a note somewhere?